### PR TITLE
log warning to file

### PR DESCRIPTION
**Pull request by Christian-B**
Optional ability to log warning to  file with just a count at the end

### CODE REVIEW

File: spinn_utilities/log.py [hunk 1]
```diff
@@ -245,10 +265,21 @@ def _atexit_handler(cls):
         messages = cls._repeat_log()
         if messages:
             level = logging.getLevelName(cls.__repeat_at_end)
-            print("\nThese log messages where generated at level {} or above"
-                  "".format(level), file=sys.stderr)
-            for message in messages:
-                print(message, file=sys.stderr)
+            if cls.__report_file:
+                if cls.__report_file:
```
Comment (andrewgait) [inline spinn_utilities/log.py:269]: Something's wrong here, the same if statement is repeated twice...

Comment (Christian-B) [reply spinn_utilities/log.py]: fixed

File: spinn_utilities/log.py [hunk 2]
```diff
@@ -174,6 +176,21 @@ def set_kill_level(cls, level=None):
...
+        :param report_file:
+        :param write_normal:
+        :return:
+        """
+        cls.__report_file = report_file
+        level = logging.getLevelName(cls.__repeat_at_end)
+        with open(report_file, "a") as report_file:
+            report_file.write(
+                "This is a record of all logged messages at level {} or "
+                "above\n".format(level))
```
Comment (andrewgait) [inline spinn_utilities/log.py:192]: Do we have easy access anywhere here to add either a job ID or a timestamp?

File: spinn_utilities/log.py [hunk 3]
```diff
@@ -245,10 +265,20 @@ def _atexit_handler(cls):
         messages = cls._repeat_log()
         if messages:
             level = logging.getLevelName(cls.__repeat_at_end)
-            print("\nThese log messages where generated at level {} or above"
-                  "".format(level), file=sys.stderr)
-            for message in messages:
-                print(message, file=sys.stderr)
+            if cls.__report_file:
+                print("\nWARNING: {} log messages where generated at "
```
Comment (andrewgait) [inline spinn_utilities/log.py:269]: "were generated"

File: spinn_utilities/log.py [hunk 4]
```diff
@@ -245,10 +265,20 @@ def _atexit_handler(cls):
...
+                      "level {} or above.".format(len(messages), level),
+                      file=sys.stderr)
+                print("This may mean that the results are invalid.",
+                      file=sys.stderr)
+                print("You are advised to check the details of these here: {}"
+                      "".format(cls.__report_file),
+                      file=sys.stderr)
+            else:
+                print("\nThese log messages where generated at level {} or "
```
Comment (andrewgait) [inline spinn_utilities/log.py:278]: "were generated"

**Review by andrewgait — CHANGES_REQUESTED** [review]
Some minor edits and a question: is it possible to add a timestamp/job ID to the warnings_log.txt file?

File: spinn_utilities/log.py (hunk 2 above)
Comment (Christian-B) [reply spinn_utilities/log.py:192]: There is no need to do that here due to the file Fron end common picks

There will be a separate report file for each run in the report director such as

reports/2020-12-15-07-44-21-617755/run_1/

File: spinn_utilities/log.py (hunk 3 above)
Comment (Christian-B) [reply spinn_utilities/log.py:269]: fixed

File: spinn_utilities/log.py (hunk 4 above)
Comment (Christian-B) [reply spinn_utilities/log.py:278]: fixed

**Review by andrewgait — APPROVED** [review]

**Pull request merged**